### PR TITLE
[Testing:Developer] Add eqeqeq eslint rule

### PR DESCRIPTION
**Pull request by williamjallen**
### Please check if the PR fulfills these requirements:

* [ ] Documentation has been updated/added if relevant

### What is the current behavior?
The Javascript `==` and `!=` operators are generally not recommended because they can behave in [unexpected ways](https://dorey.github.io/JavaScript-Equality-Table/).  Our `eslint` configuration does not currently specify which operators are preferred.  I have recently observed inconsistencies in Javascript code style relating to the equality operator while reviewing other PRs.

### What is the new behavior?
This PR adds the [`eqeqeq`](https://eslint.org/docs/latest/rules/eqeqeq) and rule to our `eslint` configuration.

A documentation PR to update [this page](https://submitty.org/developer/software_and_system_design/coding_style_guide/javascript) will be forthcoming.

### CODE REVIEW

**Comment by MasterOdin** [comment]
Could we split this into two PRs, one for `eqeqeq` and one for `space-before-blocks`? I think the latter can be merged pretty instantly as it's just style changes, but the former I think we can do some amount of code changes / longer PR review where we currently just ignore the rule.

**Comment by williamjallen** [comment]
Alright, will do.  Do you have suggestions for other good rules that should be added?  I looked through the list and didn't see anything that stood out, but I'm not familiar with what rules are considered to be common practice/useful.

**Comment by MasterOdin** [comment]
Looking at #8133, I would recommend adding [`@typescript-eslint/type-annotation-spacing`](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/type-annotation-spacing.md). As we grow the usage of modules, I would also recommend adding the `import/order` rule from [eslint-plugin-import](https://www.npmjs.com/package/eslint-plugin-import) with following definition:

```js
    'import/order': [
      'error',
      {
        alphabetize: { caseInsensitive: true, order: 'asc' },
        groups: [
          ['builtin', 'external'],
          'parent',
          'sibling',
          'index',
          'object',
        ],
        'newlines-between': 'always',
      },
    ],
```

Otherwise, not sure there's anything in particular that jumps out at me, but some amount of this is just growing the style guide stuff organically as we see people differing on style, and going "can we mandate particular style here?"

File: site/public/js/admin-team-form.js [hunk 1]
```diff
@@ -166,6 +176,7 @@ function adminTeamForm(new_team, who_id, reg_section, rot_section, user_assignme
                 break;
             }
         }
+        // eslint-disable-next-line eqeqeq
         if (found || $(this).val() == '') {
```
Comment (MasterOdin) [inline site/public/js/admin-team-form.js:180]: We could replace this with `===` only if we're confident that we're not relying on entering this if statement when `$(this).val()` is `0` or `[]` as those are both `== ''`

File: site/public/js/autograding-status.js [hunk 2]
```diff
@@ -42,10 +42,12 @@ function updateTable() {
                         new_row.insertCell().innerHTML = course_name[1];
                         new_row.insertCell().innerHTML = key2;
                         const int_box = new_row.insertCell();
+                        // eslint-disable-next-line eqeqeq
                         if (info.interactive != 0) {
```
Comment (MasterOdin) [inline site/public/js/autograding-status.js:46]: Similar to above, we can replace these only if confident that it cannot be `''` or `[]`, so probably safe to just leave these as disabled for now.

**Review by MasterOdin — CHANGES_REQUESTED** [review]

**Comment by MasterOdin** [comment]
Another rule would be [`prefer-arrow-callback`](https://eslint.org/docs/latest/rules/prefer-arrow-callback) as the standard in most code bases in my experience these days are in using arrow functions.

**Comment by williamjallen** [comment]
`prefer-arrow-callback` is [listed](https://github.com/Submitty/Submitty/blob/a350062895539e5c52e5e2ea3529d7a7f80ff704/site/.eslintrc.json#L31) in our `.eslintrc.json` file but unless I'm misunderstanding the usage of this rule, it doesn't seem to be enforced.  Example: https://github.com/Submitty/Submitty/blob/a350062895539e5c52e5e2ea3529d7a7f80ff704/site/public/js/directory.js#L11

I'll poke around in the `eslint` config some more tomorrow to figure out what's wrong.  I'll also look over the changes in this PR to resolve anything which can be done safely such as the ones requested in an earlier comment.

**Comment by MasterOdin** [comment]
Ah, it's because of the usage of `this` within the anonymous function. I missed that. Solution there would be to work to drop jQuery as a dependency I suppose.

**Comment by tkoz0** [comment]
Why is it that you are leaving quite a few `==` and `!=` as ignored? Is it because of the need to test the changes?

File: site/public/js/confetti.js [hunk 3]
```diff
@@ -16,6 +16,7 @@ function addConfetti() {
         }
     });
     window.addEventListener('keypress', (e) => {
+        // eslint-disable-next-line eqeqeq
         if (e.code === 'Enter' && canvas.style.display != 'none') {
```
Comment (MasterOdin) [inline site/public/js/confetti.js:20]:
```suggestion
        if (e.code === 'Enter' && canvas.style.display !== 'none') {
```

File: site/public/js/system-update.js [hunk 4]
```diff
@@ -140,6 +140,7 @@ function toggleAllReleases(toggleAllButton) {
     //  if toggleAllButton has collapsed - will collapse all
     //  if toggleAllButton doesn't have collapsed - will expand all
     $('.btn-toggle-release').each( (index, button) => {
+        // eslint-disable-next-line eqeqeq
         if ($(toggleAllButton).hasClass('collapsed') != $(button).hasClass('collapsed')) {
```
Comment (MasterOdin) [inline site/public/js/system-update.js:144]:
```suggestion
        if ($(toggleAllButton).hasClass('collapsed') !== $(button).hasClass('collapsed')) {
```

[`$.hasClass`](https://api.jquery.com/hasclass/) only returns a boolean.

File: site/ts/activity-dashboard.js [hunk 5]
```diff
@@ -71,12 +72,15 @@ export function comparator (row1, row2, n, dir) {
...
+    // eslint-disable-next-line eqeqeq
     if (x != '' && y == '') {
         return false;
     }
+    // eslint-disable-next-line eqeqeq
     else if (x == '' && y != '') {
         return true;
     }
+    // eslint-disable-next-line eqeqeq
     if (i == 0 || i == 8) {
```
Comment (MasterOdin) [inline site/ts/activity-dashboard.js:84]:
```suggestion
    if (i == 0 || i === 8) {
```

File: site/ts/activity-dashboard.js [hunk 6]
```diff
@@ -90,13 +94,14 @@ export function helper (x, y, i) {
             return x < y;
         }
     }
+    // eslint-disable-next-line eqeqeq
     else if ((i <= 7 && i >= 4) || i == 9) {
```
Comment (MasterOdin) [inline site/ts/activity-dashboard.js:98]:
```suggestion
    else if ((i <= 7 && i >= 4) || i === 9) {
```

File: site/ts/activity-dashboard.js [hunk 7]
```diff
@@ -209,9 +220,11 @@ export function columnOnClick(n) {
     }
 
     for (i = 0; i < 10; i++) {
+        // eslint-disable-next-line eqeqeq
         if (i != n && $(`#${i}`).children('i').hasClass('fa-angle-up')) {
```
Comment (MasterOdin) [inline site/ts/activity-dashboard.js:224]: All usages of this function pass `n` as a number.

```suggestion
        if (i !== n && $(`#${i}`).children('i').hasClass('fa-angle-up')) {
```

File: site/ts/activity-dashboard.js [hunk 8]
```diff
@@ -209,9 +220,11 @@ export function columnOnClick(n) {
...
             $(`#${i}`).children('i').removeClass('fa-angle-up');
         }
+        // eslint-disable-next-line eqeqeq
         else if (i != n && $(`#${i}`).children('i').hasClass('fa-angle-down')) {
```
Comment (MasterOdin) [inline site/ts/activity-dashboard.js:228]: Same here

```suggestion
        // eslint-disable-next-line eqeqeq
        else if (i !== n && $(`#${i}`).children('i').hasClass('fa-angle-down')) {
```

**Review by MasterOdin — CHANGES_REQUESTED** [review]

File: site/public/js/user-profile.js [hunk 9]
```diff
@@ -284,6 +284,7 @@ $(document).ready(() => {
 
                     // Check user's current time zone, give a warning message if the user's current time zone differs from systems' time-zone
                     const offset = getCurrentUTCOffset();
+                    // eslint-disable-next-line eqeqeq
                     if (response.data.utc_offset != offset) {
```
Comment (MBon1) [inline site/public/js/user-profile.js:288]: I was able to get the same result when changing the time zone in the user profile after replacing `response.data.utc_offset != offset` with `response.data.utc_offset !== offset`.

File: site/public/js/docker_interface.js [hunk 10]
```diff
@@ -31,6 +31,7 @@ function filterOnClick() {
         const this_row = $(this);
         let hide = true;
         $(this).find('.badge').each(function () {
+            // eslint-disable-next-line eqeqeq
             if ($(this).text() == this_filter) {
```
Comment (MBon1) [inline site/public/js/docker_interface.js:35]: In the jQuery documentations, `.data` can return any type.  However, in my testing, it looks like in this function (`filterOnClick`), `.data` is only returning a `string`, the selected filter's name.

File: site/public/js/forum.js [hunk 11]
```diff
@@ -95,7 +95,7 @@ function uploadImageAttachments(attachment_box) {
         const filename = target.attr('fname');
         // eslint-disable-next-line no-undef
         for (let j = 0; j < file_array[part-1].length; j++) {
-            // eslint-disable-next-line no-undef
+            // eslint-disable-next-line no-undef, eqeqeq
             if (file_array[part-1][j].name == filename) {
```
Comment (MBon1) [inline site/public/js/forum.js:99]: I was able to get the same result when uploading an attachment to a forum post after replacing `file_array[part-1][j].name == filename` with `file_array[part-1][j].name === filename`.  When getting the type of `file_array[part-1][j].name` and `filename`, `string` is returned for both.

File: site/public/js/forum.js [hunk 12]
```diff
@@ -966,6 +974,7 @@ function readThreadStatusValues() {
 
 function dynamicScrollLoadPage(element, atEnd) {
     const load_page = $(element).data(atEnd?'next_page':'prev_page');
+    // eslint-disable-next-line eqeqeq
     if (load_page == 0) {
```
Comment (MBon1) [inline site/public/js/forum.js:978]: The same behavior was produced when scrolling through the discussion forum posts after replacing `load_page == 0` with `load_page === 0`.  When getting the type of `load_page`, `number` is returned.

File: site/public/js/forum.js [hunk 13]
```diff
@@ -985,6 +994,7 @@ function dynamicScrollLoadPage(element, atEnd) {
         load_page_callback = function(content, count) {
             spinner_down.hide();
             arrow_down.before(content);
+            // eslint-disable-next-line eqeqeq
             if (count == 0) {
```
Comment (MBon1) [inline site/public/js/forum.js:998]: The same behavior was produced when scrolling through the discussion forum posts after replacing `count == 0` with `count === 0`.  When getting the type of `count`, `number` is returned.

File: site/public/js/forum.js [hunk 14]
```diff
@@ -1006,6 +1016,7 @@ function dynamicScrollLoadPage(element, atEnd) {
         load_page_callback = function(content, count) {
             spinner_up.hide();
             arrow_up.after(content);
+            // eslint-disable-next-line eqeqeq
             if (count == 0) {
```
Comment (MBon1) [inline site/public/js/forum.js:1020]: While I was unable to trigger this `else` condition, the same behavior would likely occur if `count == 0` is replaced with `count === 0` as the code here is almost identical to the preceding condition.

**Review by MBon1 — CHANGES_REQUESTED** [review]
Including my previous comments, there were a couple of places that I found using `===` instead of `==` could work.  The files include `forum.js`, `docker_interface.js`, and `user-profile.js`.

**Comment by williamjallen** [comment]
> Why is it that you are leaving quite a few `==` and `!=` as ignored? Is it because of the need to test the changes?

Yes.  This PR actually turned out to not be as bad as I expected, but many linting PRs are difficult to test so we usually ignore any pre-existing errors which can't be corrected easily.  In some cases such as the previous requested changes, we can be certain that the change won't break anything.  In other cases, it is sometimes difficult to determine what type a variable might be so we simply ignore the changes.

**Review by MasterOdin — APPROVED** [review]
Yeah, I suspect we could resolve each of these by going through one by one, but in the sake of progress, and not letting more of these errors enter the codebase, I would say merge this now, and the remaining ignored errors can be slowly cleaned up as people touch the containing code, or some entrepreneurial soul wishes to clean it up separately.

**Review by MBon1 — APPROVED** [review]

**Pull request merged**